dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java

```diff
@@ -580,6 +571,49 @@ else if ( params.isOrganisationUnitMode( OrganisationUnitSelectionMode.CHILDREN
         return sql;
     }
 
+    /**
+     * Creates a SQL statement of descendants org units. When there are multiple
+     * levels the "or" operator will be used as junction.
+     *
+     * @param orgUnitAlias
+     * @param dimensionalItems
+     * @param helper
+     * @return a SQL statement in the format "where/and (ax."uidlevel0" in
+     *         ('orgUid-1', 'orgUid-2', 'orgUid-2') )"
+     */
+    private String descendantsOrgUnitStatement( final String orgUnitAlias,
+        final List<DimensionalItemObject> dimensionalItems, final SqlHelper helper )
+    {
+        final StringBuilder statement = new StringBuilder();
+        final Map<String, String> orgUnitColsAndUnitUids = new HashMap<>();
+
+        statement.append( helper.whereAnd() + " (" );
+
+        for ( final DimensionalItemObject itemObject : dimensionalItems )
+        {
+            final OrganisationUnit unit = (OrganisationUnit) itemObject;
+
+            final String orgUnitCol = quote( orgUnitAlias, "uidlevel" + unit.getLevel() );
+
+            if ( isNotBlank( orgUnitColsAndUnitUids.get( orgUnitCol ) ) )
+            {
+                orgUnitColsAndUnitUids.put( orgUnitCol,
+                    trimToEmpty( orgUnitColsAndUnitUids.get( orgUnitCol ) ) + ", " + wrap( unit.getUid(), "'" ) );
```

Comment:
is it not possible to use stream api here? for example collect all levels (distinct), join all related uids (Collectors.joining)...

Reply:
I would argue stream syntax should not necessarily be applied because "we can", it has a minor performance overhead and often makes code less readable IMHO.